xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,201 @@
+	// do nothing.
+	if rootClusterName != ch.rootClusterName {
+		ch.root.delete()
+		ch.root = createClusterNode(rootClusterName, ch.xdsClient, ch)
+		ch.rootClusterName = rootClusterName
+	}
+}
+
+
+// This function tries to construct a cluster update to send to CDS.
+func (ch *clusterHandler) constructClusterUpdate() {
+	// If there was an error received no op, as this simply means one of the children hasn't received an update yet.
+	if clusterUpdate, err := ch.root.constructClusterUpdate(); err == nil {
+		// For a ClusterUpdate, the only update CDS cares about is the most recent one, so opportunistically drain the
+		// update channel before sending the new update.
+		select {
+		case <-ch.updateChannel:
+		default:
+		}
+		ch.updateChannel<-clusterHandlerUpdate{chu: clusterUpdate, err: nil}
+	}
+}
+
+// This logically represents a cluster. This handles all the logic for starting and stopping
+// a cluster watch, handling any updates, and constructing a list recursively for the ClusterHandler.
+type clusterNode struct {
+	// A way to cancel the watch for the cluster.
+	cancelFunc func()
+
+	// A list of children, as the Node can be an aggregate Cluster.
+	children []*clusterNode
+
+	// A ClusterUpdate in order to build a list of cluster updates for CDS to send down to child
+	// XdsClusterResolverLoadBalancingPolicy.
+	clusterUpdate xdsclient.ClusterUpdate
+
+	// This boolean determines whether this Node has received an update or not. This isn't the best practice,
+	// but this will protect a list of Cluster Updates from being constructed if a cluster in the tree has not received
+	// an update yet.
+	receivedUpdate bool
+
+	clusterHandler *clusterHandler
+
+}
+
+// CreateClusterNode creates a cluster node from a given clusterName. This will also start the watch for that cluster.
+func createClusterNode(clusterName string, xdsClient xdsClientInterface, topLevelHandler *clusterHandler) *clusterNode {
+	print("|createClusterNode() called on ", clusterName, "|")
+	c := &clusterNode{
+		clusterHandler: topLevelHandler,
+	}
+	print("About to communicate with the xds client for:", clusterName, "|")
+	// Communicate with the xds client here.
+	c.cancelFunc = xdsClient.WatchCluster(clusterName, c.handleResp)
+	return c
+}
+
+// This function cancels the cluster watch on the cluster and all of it's children.
+func (c *clusterNode) delete() {
+	c.cancelFunc()
+	for _, child := range c.children {
+		child.delete()
+	}
+}
+
+// Construct cluster update (potentially a list of ClusterUpdates) for a node.
+func (c *clusterNode) constructClusterUpdate() ([]xdsclient.ClusterUpdate, error) { // THIS LOGIC IS WRONG
+	print("Tries constructing a cluster update for Cluster ", c.clusterUpdate.ServiceName, ". ")
+	// If the cluster has not yet received an update, the cluster update is not yet ready.
+	if !c.receivedUpdate {
+		return nil, errors.New("Tried to construct a cluster update on a cluster that has not received an update.")
```

Comment:
Nit: make a top level err for this, and reuse.

Reply:
Changed to errNotReceivedUpdate. Didn't add comment because error text seems self explanatory.